### PR TITLE
[web] Use nginx to proxy pass the api

### LINKED ISSUE #345: [spec] Use a proxy service to form a transparent communication endpoint
* Status:
* Deciders:
* Date: 2021-04-22

## Context and Problem Statement

We want to simplify the deployment and operations of a Monocle service.
How do we define the network topology and monitoring of the components?

## Considered Options

* Docker compose healthcheck and ports
* Nginx
* Kubernetes ingress
* HAProxy
* Envoy

## Decision Outcome

Chosen option: ""

### Positive Consequences

* …

### Negative Consequences

* …

### Docker compose

* Good, because it is a well known service composition.
* Bad, because the services share a single network and the configuration is coupled with the ip addresses.

## Nginx

* Good, because it is simple.

### Kubernetes

* Good, because production grade container orchestration.
* Bad, because it is complex.

### HAProxy

* Good, because it is reliable.

### Envoy

* Good, because it is designed for microservices.
* Good, because it has a mature ecosystem, including logging, metrics and gRPC.
* Bad, because it requires a complex configuration, e.g. for dynamic configuration.

### PR DESCRIPTION
**Pull request by TristanCacqueray**
This change simplifies deployment and enable more flexibility to
setup new api service:

- Add service port configuration to enable single host multi deployment
- Consolidate MONOCLE_URL into MONOCLE_PUBLIC_URL
- Remove CORS default configuration

This change also enables podman v3:
- Add missing compose healthcheck attribute
- Move the web dockerfile next to the compose file

Finally this change also improves the compose.dhall file by adding
helper functions.

Fixes #345

### CODE REVIEW

**Review by morucci — CHANGES_REQUESTED** [review]
Monocle web does not start in the start-with-podman script:
[fabien@ceres monocle]$ podman logs -f monocle_web
+ test -z ''
+ test -z 'Monocle Dev Podman deployment'
+ sed -i -e 's@__TITLE__@Monocle Dev Podman deployment@' /app/index.html
+ exec nginx -g 'daemon off;'
nginx: [emerg] host not found in upstream "api" in /etc/nginx/conf.d/monocle.conf:11

File: .env.sample [hunk 1]
```diff
@@ -1,10 +1,13 @@
 ### Here you can overwrite default values ###
 ### NOTE: If you are using Monocle on external server, replace localhost in
 ###       MONOCLE_URL and MONOCLE_API_URL with your public ip address.
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply .env.sample:3]: Done

File: README.md [hunk 2]
```diff
@@ -177,10 +177,9 @@ For a local deployment, default settings are fine.
 
 The following settings are available in the `.env` file:
 
-- `MONOCLE_URL=<host or ip>` to configure the URL serving the Web UI
-  (default `http://localhost:3000`).
-- `MONOCLE_API_URL=<host or ip>` to configure the URL serving the API
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply README.md:182]: Oops indead, we can drop it from the package. That's only needed to run the javascript development server with a custom api url.

File: contrib/start-with-podman.sh [hunk 3]
```diff
@@ -26,18 +26,19 @@
...
 # podman restart monocle_api monocle_crawler
 #
 # To hack the WEB UI (after a complete run with "start") run
 # the script with the "start-web-dev" parameter.
+# -> firefox http://localhost:3000
 
 set -x
 
-export PUBLIC_ADDRESS=localhost
+export PUBLIC_ADDRESS=/
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply contrib/start-with-podman.sh:41]: Done.

File: docker-compose.dhall [hunk 4]
```diff
@@ -91,21 +106,14 @@ let createApiService =
                         , mapValue = "/etc/monocle/config.yaml"
                         }
                       , { mapKey = "ELASTIC_CONN", mapValue = "elastic:9200" }
-                      , { mapKey = "ALLOW_ORIGIN"
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply docker-compose.dhall:94]: Maybe that could be useful to disable cors?

File: contrib/start-with-podman.sh [hunk 5]
```diff
@@ -81,7 +80,8 @@ if [ "$1" == "create" ]; then
                --pod monocle \
                --add-host monocle_api:127.0.0.1 \
                --add-host monocle_elastic:127.0.0.1 \
-               monocle_web bash -c "set -x; sed -i -e \"s@__API_URL__@http://$PUBLIC_ADDRESS:9876@\" build/index.html; sed -i -e \"s@__TITLE__@Monocle Dev Podman deployment@\" build/index.html; serve -s /app/ -l 3000"
+               -e REACT_APP_TITLE="Monocle Dev Podman deployment"
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply contrib/start-with-podman.sh:83]: Oops, sorry, I didn't have enough disk space to both test podman and docker-compose.

**Review by morucci — APPROVED** [review]

**Comment by TristanCacqueray** [comment]
:+1:

**Pull request merged**